linkerd/proxy/http/src/detect.rs

```diff
@@ -0,0 +1,227 @@
+use crate::Version;
+use bytes::BytesMut;
+use linkerd2_error::Error;
+use linkerd2_io::{self as io, AsyncReadExt};
+use linkerd2_proxy_transport::{Detect, NewDetectService};
+use linkerd2_stack::layer;
+use tokio::time;
+use tracing::{debug, trace};
+
+const H2_PREFACE: &'static [u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+const H2_FIRST_LINE_LEN: usize = 16;
+
+#[inline]
+fn is_h2_first_line(line: &[u8]) -> bool {
+    line.len() == H2_FIRST_LINE_LEN && line == &H2_PREFACE[..H2_FIRST_LINE_LEN]
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct DetectHttp(());
+
+impl DetectHttp {
+    pub fn layer<N>(
+        timeout: time::Duration,
+    ) -> impl layer::Layer<N, Service = NewDetectService<N, Self>> + Clone {
+        NewDetectService::layer(timeout, Self(()))
+    }
+}
+
+#[async_trait::async_trait]
+impl Detect for DetectHttp {
+    type Protocol = Version;
+
+    async fn detect<I: io::AsyncRead + Send + Unpin + 'static>(
+        &self,
+        io: &mut I,
+        buf: &mut BytesMut,
+    ) -> Result<Option<Version>, Error> {
+        let mut scan_idx = 0;
+        let mut maybe_h1 = true;
+        let mut maybe_h2 = true;
+
+        loop {
+            // Read data from the socket or timeout detection.
+            trace!(
+                capacity = buf.capacity(),
+                scan_idx,
+                maybe_h1,
+                maybe_h2,
+                "Reading"
+            );
+            let sz = io.read_buf(buf).await?;
+            if sz == 0 {
+                // No data was read because the socket closed or the
+                // buffer capacity was exhausted.
+                debug!(read = buf.len(), "Could not detect protocol");
+                return Ok(None);
+            }
+
+            // HTTP/2 checking is faster because it's a simple string match. If
+            // we have enough data, check it first. In almost all cases, the
+            // whole preface should be available from the first read.
+            if maybe_h2 && buf.len() >= H2_PREFACE.len() {
+                trace!("Checking H2 preface");
+                if &buf[..H2_PREFACE.len()] == H2_PREFACE {
+                    trace!("Matched HTTP/2 prefix");
+                    return Ok(Some(Version::H2));
+                }
+
+                // Not a match. Don't check for an HTTP/2 preface again.
+                maybe_h2 = false;
+            }
+
+            if maybe_h1 {
+                // Scan up to the first line ending to determine whether the
+                // request is HTTP/1.1.
+                for i in scan_idx..(buf.len() - 1) {
+                    if &buf[i..=i + 1] == b"\r\n" {
```

Comment:
nit, take it or leave it: I think it's possible that this logic could also be accomplished with [`std::slice::windows`](https://doc.rust-lang.org/std/primitive.slice.html#method.windows), which might be a little bit neater:

```rust
for window in &buf[scan_index..].windows() {
    if window == b"\r\n" {
        // ...
    }
}
```